File: MatlabOctaveCodeSnippets/snip_systems_s_to_z_plane.m
close all
chebyshevFilterType=1; %1 has ripple in passband, 2 in stopband
filterOrder=4; %H(s) filter order
zTosMapping='s=z*Fs'; %some transformation
%zTosMapping='s=(2*Fs*(z-1))./(z+1)'; %bilinear transformation
if chebyshevFilterType==1
    wfilter=8; %passband frequency (rad/s);
    passbandRipple=1; %maximum ripple in passband (dB)
    [Bs,As]=cheby1(filterOrder,passbandRipple,wfilter,'s'); %H(s)
elseif chebyshevFilterType==2
    wfilter=15; %stopband frequency (rad/s);
    stopbandRipple=40; %ripple down from the peak passband value (dB)
    [Bs,As]=cheby2(filterOrder,stopbandRipple,wfilter,'s'); % H(s)
end
Fs=(20*wfilter)/(2*pi); %choose sampling frequency (Hz)
%% Plot H(s) and the corresponding H(z)
slim=20;
[Xs,Ys]=meshgrid(linspace(-slim,slim,100)); %100x100 grid in S plane
s=Xs+1j*Ys; %matrix with all points of interest in s plane
Hs=polyval(Bs,s)./polyval(As,s); %H(s)
zlim=10;
[Xz,Yz]=meshgrid(linspace(-zlim,zlim,100)/Fs); %100x100 grid in Z plane
z=Xz+1j*Yz; %matrix with all points of interest in z plane
%for each point z at Z plane, find s and the corresponding value H(s)
%s=(2*Fs*(z-1))./(z+1); %use bilinear transformation
eval(zTosMapping); %for each z, find the corresponding s
%Use H(z)=H(s)|s=2Fs(z-1)/(z+1)
Hz=polyval(Bs,s)./polyval(As,s); %H(z)
figure(1)
meshc(Xs,Ys,20*log10(abs(Hs))); %plot in dB
xlabel('Re\{s\} (\sigma)'); ylabel('Im\{s\} (j\omega)');
zlabel('20 log_{10} |H(s)|');
figure(2)
meshc(Xz,Yz,20*log10(abs(Hz))); %plot in dB
xlabel('Re\{z\}'); ylabel('Im\{z\}');
zlabel('20 log_{10} |H(z)|');

%% The frequency responses
figure(3)
s=1j*linspace(-slim,slim,100); %w varies from -slim,slim
Hs_w=polyval(Bs,s)./polyval(As,s); %H(s) is in fact H(jw) here
subplot(311)
plot(imag(s),20*log10(abs(Hs_w)))
xlabel('\omega (rad/s)'), ylabel('20 log_{10} |H(j\omega)|');
axis tight
%find the mapping from s to Z over the imaginary axis of Z
z=1j*linspace(-pi,pi,100); %w varies from -slim,slim
eval(zTosMapping); %for each z, find the corresponding s
Hz_new_w=polyval(Bs,s)./polyval(As,s); %H(z) over Imaginary{Z}
subplot(312)
plot(imag(z),20*log10(abs(Hz_new_w)))
xlabel('Im\{z\}'), ylabel('20 log_{10} |H(Im\{z\})|');
axis tight
%find the mapping from s to Z over the unit circle
z=exp(1j*linspace(-pi,pi,100)); %w varies from -pi,pi
%s=(2*Fs*(z-1))./(z+1); %use bilinear transformation
eval(zTosMapping);
Hz_W=polyval(Bs,s)./polyval(As,s); %H(z) over the unit circle
subplot(313)
plot(angle(z),20*log10(abs(Hz_W)))
xlabel('\Omega (rad)'), ylabel('20 log_{10} |H(e^{j\Omega})|');
axis tight

File: PythonCodeSnippets/snip_systems_s_to_z_plane.py
import numpy as np
import matplotlib.pyplot as plt
from scipy import signal

# Set parameters
chebyshev_filter_type = 1  # 1 has ripple in passband, 2 in stopband
filter_order = 4  # H(s) filter order
z_to_s_mapping = "s = z * Fs"  # some transformation

if chebyshev_filter_type == 1:
    w_filter = 8  # passband frequency (rad/s)
    passband_ripple = 1  # maximum ripple in passband (dB)
    Bs, As = signal.cheby1(filter_order, passband_ripple, w_filter, analog=1)
elif chebyshev_filter_type == 2:
    w_filter = 15  # stopband frequency (rad/s)
    stopband_ripple = 40  # ripple down from the peak passband value (dB)
    Bs, As = signal.cheby2(filter_order, stopband_ripple, w_filter, analog=1)

Fs = (20 * w_filter) / (2 * np.pi)  # choose sampling frequency (Hz)

# Plot H(s) and the corresponding H(z)
slim = 20
aux1 = np.linspace(-slim, slim, 100)
Xs, Ys = np.meshgrid(aux1, aux1)
s = Xs + 1j * Ys
Hs = np.divide(np.polyval(Bs, s), np.polyval(As, s))

zlim = 10
Xz, Yz = np.meshgrid(
    np.linspace(-zlim, zlim, 100) / Fs, np.linspace(-zlim, zlim, 100) / Fs
)
z = Xz + 1j * Yz

# for each point z at Z plane, find s and the corresponding value H(s)
# s = (2 * Fs * (z - 1)) / (z + 1)  # use bilinear transformation
exec(z_to_s_mapping)  # for each z, find the corresponding s
Hz = np.divide(np.polyval(Bs, s), np.polyval(As, s))

# Plot in dB
fig1 = plt.figure(1)
ax1 = fig1.add_subplot(111, projection="3d")
ax1.plot_surface(Xs, Ys, 20 * np.log10(np.abs(Hs)), cmap="viridis")
ax1.set_xlabel("Re{s} (σ)")
ax1.set_ylabel("Im{s} (jω)")
ax1.set_zlabel("20 log10 |H(s)|")

fig2 = plt.figure(2)
ax2 = fig2.add_subplot(111, projection="3d")
ax2.plot_surface(Xz, Yz, 20 * np.log10(np.abs(Hz)), cmap="viridis")
ax2.set_xlabel("Re{z}")
ax2.set_ylabel("Im{z}")
ax2.set_zlabel("20 log10 |H(z)|")

# The frequency responses
fig3, (ax3, ax4, ax5) = plt.subplots(3, 1, figsize=(8, 12))

# H(s) is in fact H(jw) here
s_w = 1j * np.linspace(-slim, slim, 100)
Hs_w = np.divide(np.polyval(Bs, s_w), np.polyval(As, s_w))
ax3.plot(np.imag(s_w), 20 * np.log10(np.abs(Hs_w)))
ax3.set_xlabel("ω (rad/s)")
ax3.set_ylabel("20 log10 |H(jω)|")

# find the mapping from s to Z over the imaginary axis of Z
z_w = 1j * np.linspace(-np.pi, np.pi, 100)
exec(z_to_s_mapping)
Hz_new_w = np.divide(np.polyval(Bs, s), np.polyval(As, s))
ax4.plot(np.imag(z_w), 20 * np.log10(np.abs(Hz_new_w)))
ax4.set_xlabel("Im{z}")
ax4.set_ylabel("20 log10 |H(Im{z})|")

# find the mapping from s to Z over the unit circle
z_W = np.exp(1j * np.linspace(-np.pi, np.pi, 100))
exec(z_to_s_mapping)
Hz_W = np.divide(np.polyval(Bs, s), np.polyval(As, s))
ax5.plot(np.angle(z_W), 20 * np.log10(np.abs(Hz_W)))
ax5.set_xlabel("Ω (rad)")
ax5.set_ylabel("20 log10 |H(e^{jΩ})|")

plt.tight_layout()
plt.show()
